### ISSUE TITLE
New Infrastructure - Talking to the BiocondaBot :)

### ISSUE THREAD

**Issue by epruesse** [bioconda/bioconda-recipes#14400]
Hey everyone,

as you may have noticed, @BiocondaBot is now at your service. So far, it does only a few things for us:

- **Linting Recipe(s)**: this is an additional (not mandatory) check of your recipes for known problems. The checks are not new, but the bot is faster than CircleCI and will usually get back to you within 20 seconds, telling you where your recipes had problems and what you should do to repair them.

- **Checking for Artifacts**: as your builds on CircleCI succeed, the bot will pick up the packages that were created and make a post with links for you to download or install the preview package for additional testing and simplified debugging. Watch out if it found nothing - your recipe might not have been building anything. 

- **Commands**: The bot supports a few commands mostly designed to kick things off again if the events from Github got lost:
  - `@biocondabot hello`: To check if the bot is listening, shout out to it and it will answer.
  - `@biocondabot recheck`: (Re-)triggers **Linting Recipe(s)** and **Checking for Artifacts**.
  - `@biocondabot rebuild`: Triggers a (re)build on CircleCI. Use to repair cases where the **bioconda-tests** CircleCI workflow wasn't queued and the dot stays orange forever.

**Comment by HenningTimm** [comment]
Very cool :) Just to check that I understand this right: I just mention the bot like @bioconda-bot hello and it will perform the actions shown above? When I do so in a branch, it will notice that and only rebuild my branch, correct?

**Comment by epruesse** [comment]
Yes, when you @Mention it from a PR, @BiocondaBot will know which issue you want it to work on. You can also talk to it via Gitter, but in that case must add the PR number (`@BiocondaBot rebuild #123`) and should be careful to get the right number, lest you mess with someone else's PR.

It's not very smart - you have to put the @Mention at the beginning of a line and use the exact command word. I didn't have time to add any fancy NLP code. As always, PRs arewelcome.

**Comment by jenzopr** [comment]
Nice! What about the autobump feature? How can I set this up for one of my recipes?

**Comment by epruesse** [comment]
@jenzopr I hear you, and I wish I saw a way to have that work.

Very briefly: the memory and disk space needs for calling the `conda-build` API to determine whether a package needs a pinning bump are beyond what we can run on our frugal little AWS instance. We'd need a 16GB box with 50GB disk to run that, and would look at $100+ per month, which we don't currently have funding for. Getting the requirements for Python packages also requires running `setup.py` with modified Python base libs, for which we would need an S2 or BSL-3 certified lab.

I'm running it manually myself right now in small batches. Once we have worked our way through the backlog, it can hopefully go back on nightly mode and check for updates to all packages once a day. You could in theory also run it yourself as well, it's part of `bioconda-utils`.

What I would *really really* appreciate help with is working on all those PRs. If you have *any* feedback on how we might better motivate and honor the work people contribute to this end, I'm all ear! We need you all!

Note on the side:
**The community is the important part, not the infrastructure!** Wikipedia, back when it was founded, was considered having "poor technology". It's competitors (yes it had those!) had more advanced software, and focused more on that. They failed. For a community project, **tech is secondary. It's all about the people.**

**Comment by jenzopr** [comment]
Hey Elmar,
your remarks are understandable and I can imagine that running the autobump requires a lot of CPU time. We'll push the updates ourselves, that's not a big deal.

Concerning your cry for help: Let me see what I can do. I guess it's not about one or two persons stepping in, but the community as a whole has to be either more active (which means, in my limited words, to manually pick up a PR, check the error on CircleCI, come up with a solution and push commits, start again) or the process of helping out has to be much faster. Did I pick this up correctly?

**Comment by epruesse** [comment]
@jenzopr - Would you consider working on the docs for this? You can PM me with questions on gitter. You can run `bioconda-utils autobump` locally to update your own packages.

And yes, helping with the generated PRs means exactly that. Usually, what I do is
- Check if it's the lint. If so, see if I can fix that.
- Check if it's failing on the CI. Look at the log and get the error. Very often, it's `g++ not found` or something like that. Sometimes it's missing requirements.
- Tweak the `build.sh`.
- Repeat until cooked.

Partially it might be that there only very few of us who look at the PRs in general, helping out, not just focusing on their own.

**Comment by epruesse** [comment]
@kepbod Any reason why you unpinned this issue?

**Comment by kepbod** [comment]
@epruesse I am sorry. I have no idea what's going on. Maybe I clicked some button for wrong to unpin this issue. I didn't mean it.

**Comment by epruesse** [comment]
No worries. Just wondering.

**Comment by nfahlgren** [comment]
This is really cool @epruesse! Based on #15332 should automated autobump PRs be reviewed by a recipe's maintainer(s) (or anyone else of course)? 

Happy to help review PRs more generally too!

**Comment by epruesse** [comment]
TL'DR: We don't have a special policy as yet. Use your own judgement for now.

The general consensus leading up to 15332 was to just slightly raise the bar to prevent accidents caused by rushed recipes. Not every commit / PR is the same though. New recipes are more difficult than updates, and minor patch levels without changes to build or dependencies are often trivial. We didn't "model" any of that, hoping that a simple rule would suffice.

There is also the issue that Github does not even allow us to differentiate between PRs generated by machine accounts (such as BiocondaBot) and those generated by actual people. So we can't demand anything special for those PRs without disabling merges for everyone and doing our own policy verification via the bot. At this time, we can't do that because we don't have the resources to ensure that the bot's hosting will be available at all times.


Since it's a lot of work to stay on top of the changes required to build every released update for the recipes in our collection, I personally don't think we need to put additional constraints on people willing to help. If you can get an Autobump PR to build and pass tests cleanly and you think the updated package is doing fine, go ahead and merge it! Request help as you need though.

**Comment by nfahlgren** [comment]
Thanks @epruesse, makes sense!
